File: src/python/filter_ics/fetch.py
"""Fetch the source iCalendar feed from a URL."""

from __future__ import annotations

import httpx


class FetchError(Exception):
    """Raised on network error, timeout, or non-2xx response (exit 1)."""


def fetch_calendar(url: str, timeout: float = 30.0) -> bytes:
    """Return the raw body of the source feed.

    Raises :class:`FetchError` on any transport error or non-2xx status.
    """
    try:
        response = httpx.get(url, timeout=timeout, follow_redirects=True)
    except httpx.HTTPError as exc:
        raise FetchError(f"could not fetch source feed: {exc}") from exc

    if response.status_code // 100 != 2:
        raise FetchError(f"source feed returned HTTP {response.status_code}")

    return response.content

File: src/dotnet/FilterIcs/Fetch.cs
namespace FilterIcs;

/// <summary>Raised on network error, timeout, or non-2xx response (exit 1).</summary>
public sealed class FetchException : Exception
{
    public FetchException(string message, Exception? inner = null) : base(message, inner) { }
}

public static class Fetch
{
    /// <summary>Fetch the source feed body as text. Throws <see cref="FetchException"/> on failure.</summary>
    public static async Task<string> FetchCalendarAsync(string url, TimeSpan? timeout = null)
    {
        using var client = new HttpClient { Timeout = timeout ?? TimeSpan.FromSeconds(30) };
        HttpResponseMessage response;
        try
        {
            response = await client.GetAsync(url);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new FetchException($"could not fetch source feed: {ex.Message}", ex);
        }

        if ((int)response.StatusCode / 100 != 2)
            throw new FetchException($"source feed returned HTTP {(int)response.StatusCode}");

        return await response.Content.ReadAsStringAsync();
    }
}
